File: liqp-core/src/test/java/liqp/nodes/OutputNodeTest.java
package liqp.nodes;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

import junitparams.JUnitParamsRunner;
import junitparams.Parameters;
import liqp.LiquidParser;
import liqp.Template;
import org.antlr.runtime.RecognitionException;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(JUnitParamsRunner.class)
public class OutputNodeTest {

  public static String[] keywordParams() {
    return new String[]{
          "capture",
          "endcapture",
          "comment",
          "endcomment",
          "raw",
          "endraw",
          "if",
          "elsif",
          "endif",
          "unless",
          "endunless",
          "else",
          "contains",
          "case",
          "endcase",
          "when",
          "cycle",
          "for",
          "endfor",
          "in",
          "and",
          "or",
          "tablerow",
          "endtablerow",
          "assign",
          "include",
          "with",
          "end",
          "break",
          "continue",
    };
  }

  public static String[][] badKeywordVariables() {
    return new String[][]{
          {"true", "true"},
          {"false", "false"},
          {"nil", ""},
          {"null", ""},
          // {"empty", "Object"},
    };
  }

  @Test
  public void applyTest() throws RecognitionException {

    String[][] tests = {
          {"{{ X }}", "mu"},
          {"{{ 'a.b.c' | split:'.' | first | upcase }}", "A"},
    };

    for (String[] test : tests) {

      Template template = LiquidParser.newInstance().parse(test[0]);
      String rendered = String.valueOf(template.render("{\"X\" : \"mu\"}"));

      assertThat(rendered, is(test[1]));
    }
  }

  @Test
  @Parameters(method = "keywordParams")
  public void testKeywordsAsVariables(String keyword) {

    String test = "{{" + keyword + "}}";
    String expected = keyword + "_" + Integer.toString(keyword.length());
    String json = "{\"" + keyword + "\" : \"" + expected + "\" }";
    Template template = LiquidParser.newInstance().parse(test);
    String rendered = template.render(json);

    assertThat(rendered, is(expected));
  }

  @Test
  @Parameters(method = "badKeywordVariables")
  public void badKeywordAsVariableTest(String keyword, String expected) {
    String test = "{{" + keyword + "}}";
    String json = "{\"" + keyword + "\" : \"bad\" }";
    Template template = LiquidParser.newInstance().parse(test);
    String rendered = template.render(json);

    assertThat(rendered, is(expected));
  }
}

File: liqp-core/src/test/java/liqp/nodes/OutputNodeTest.kt
package liqp.nodes

import assertk.assert
import assertk.assertions.isEqualTo
import liqp.LiquidParser
import liqp.createTestParser
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

@RunWith(Parameterized::class)
class OutputNodeTest(val template: String, val expect: String) {
  @Test
  fun run() {

    val template = createTestParser{}.parse(template)
    val rendered = template.renderJson("{\"X\" : \"mu\"}")

    assert(rendered).isEqualTo(expect)
  }

  companion object {
    @JvmStatic @Parameterized.Parameters
    fun params() = arrayOf(arrayOf("{{ X }}", "mu"), arrayOf("{{ 'a.b.c' | split:'.' | first | upcase }}", "A"))
  }
}

@RunWith(Parameterized::class)
class OutputNodeKeywordTest(val keyword:String) {
  @Test
  fun testKeywordsAsVariables() {

    val test = "{{$keyword}}"
    val expected = keyword + "_" + Integer.toString(keyword.length)
    val json = "{\"$keyword\" : \"$expected\" }"
    val template = createTestParser {  }.parse(test)
    val rendered = template.renderJson(json)

    assert(rendered).isEqualTo(expected)
  }

  companion object {
    @JvmStatic @Parameterized.Parameters(name = "{0}")
    fun keywordParams(): Array<String> {
      return arrayOf("capture", "endcapture", "comment", "endcomment", "raw", "endraw", "if", "elsif",
          "endif", "unless", "endunless", "else", "contains", "case", "endcase", "when", "cycle",
          "for", "endfor", "in", "and", "or", "tablerow", "endtablerow", "assign", "include",
          "with", "end", "break", "continue")
    }
  }
}

@RunWith(Parameterized::class)
class OutputNodeBadKeywordTest(val keyword: String, val expected: String) {

  @Test fun badKeywordAsVariableTest() {
    val test = "{{$keyword}}"
    val json = "{\"$keyword\" : \"bad\" }"
    val template = createTestParser { }.parse(test)
    val rendered = template.renderJson(json)

    assert(rendered).isEqualTo(expected)
  }

  companion object {

    @JvmStatic @Parameterized.Parameters
    fun badKeywordVariables(): Array<Array<String>> {
      return arrayOf(arrayOf("true", "true"), arrayOf("false", "false"), arrayOf("nil", ""), arrayOf("null", "")) // {"empty", "Object"},
    }
  }
}
